keeper/component/dispatcher/state.go

```diff
@@ -0,0 +1,504 @@
+	if err != nil {
+		d.logger.Error(
+			"error in dispatched amounts walking by source protocol ID",
+			"error",
+			err,
+		)
+
+		return []*dispatchertypes.DispatchedAmountEntry{}
+	}
+
+	return amounts
+}
+
+func (d *Dispatcher) GetDispatchedAmountsByDestinationProtocolID(
+	ctx context.Context,
+	protocolID core.ProtocolID,
+) []*dispatchertypes.DispatchedAmountEntry {
+	amounts := []*dispatchertypes.DispatchedAmountEntry{}
+
+	rng := collections.NewPrefixedPairRange[int32, DispatchedAmountsKey](int32(protocolID))
+
+	err := d.dispatchedAmounts.Indexes.ByDestinationProtocolID.Walk(
+		ctx,
+		rng,
+		func(_ int32, k DispatchedAmountsKey) (stop bool, err error) {
+			entry, err := d.getDispatchedAmountEntryFromKey(ctx, k)
+			if err != nil {
+				return true, err
+			}
+
+			amounts = append(amounts, &entry)
+
+			return false, nil
+		},
+	)
+	if err != nil {
+		d.logger.Error(
+			"error in dispatched amounts walking by destination protocol ID index",
+			"error",
+			err,
+		)
+
+		return []*dispatchertypes.DispatchedAmountEntry{}
+	}
+
+	return amounts
+}
+
+func (d *Dispatcher) GetAllDispatchedAmounts(
+	ctx context.Context,
+) []dispatchertypes.DispatchedAmountEntry {
+	amounts := []dispatchertypes.DispatchedAmountEntry{}
+
+	err := d.dispatchedAmounts.Walk(
+		ctx,
+		nil,
+		func(k DispatchedAmountsKey, v dispatchertypes.AmountDispatched) (stop bool, err error) {
+			entry, err := d.getDispatchedAmountEntryFromKey(ctx, k)
+			if err != nil {
+				return true, err
+			}
+
+			amounts = append(amounts, entry)
+
+			return false, nil
+		},
+	)
+	if err != nil {
+		d.logger.Error("error in dispatched amounts walking all values")
+
+		return []dispatchertypes.DispatchedAmountEntry{}
+	}
+
+	return amounts
+}
+
+func (d *Dispatcher) getDispatchedAmountEntryFromKey(
+	ctx context.Context,
+	k DispatchedAmountsKey,
+) (dispatchertypes.DispatchedAmountEntry, error) {
+	var entry dispatchertypes.DispatchedAmountEntry
+
+	value, err := d.dispatchedAmounts.Get(ctx, k)
+	if err != nil {
+		return entry, errorsmod.Wrap(err, "failed to get dispatched amount")
+	}
+
+	sourceID, err := core.NewCrossChainID(core.ProtocolID(k.K1()), k.K2())
+	if err != nil {
+		return entry, errorsmod.Wrap(err, "failed to create source cross-chain ID")
+	}
+
+	destID, err := core.ParseCrossChainID(k.K3())
+	if err != nil {
+		return entry, errorsmod.Wrap(err, "failed to create destination cross-chain ID")
+	}
+
+	entry.SourceId = &sourceID
+	entry.DestinationId = &destID
+	entry.Denom = k.K4()
+	entry.AmountDispatched = value
+
+	return entry, nil
+}
+
+// ====================================================================================================
+// Dispatched Counts
+// ====================================================================================================
+
+// DispatchedCountsKey is defined as:
+// (source protocol ID, source counterparty ID, destination cross-chain ID).
+type DispatchedCountsKey = collections.Quad[int32, string, int32, string]
```

Comment:
nit, this is a `QuadKey` so the comment is out-of-date

```suggestion
// DispatchedCountsKey is defined as:
// (source protocol ID, source counterparty ID, destination protocol ID, destination counterparty ID ID).
type DispatchedCountsKey = collections.Quad[int32, string, int32, string]
```